v0/destinations/one_signal/transform.js

```diff
@@ -0,0 +1,275 @@
+    devicePayload.device_type = integrationsObj.deviceType;
+    devicePayload.identifier = integrationsObj.identifier;
+  }
+};
+const getExternalIdentifiersMapping = message => {
+  // Data Structure expected:
+  // context.externalId: [ {type: playerId, id: __id}]
+  const externalId = get(message, "context.externalId");
+  let playerId;
+  if (externalId && Array.isArray(externalId)) {
+    externalId.forEach(id => {
+      if (id.type === "playerId") {
+        playerId = getDestinationExternalID(message, id.type);
+      }
+    });
+  }
+  return playerId;
+};
+
+const identifyResponseBuilder = (message, { Config }) => {
+  const { appId, emailDeviceType, smsDeviceType } = Config;
+
+  if (!appId) {
+    throw new CustomError(
+      "[OneSignal]: appId is required for creating/adding a device",
+      400
+    );
+  }
+
+  let { endpoint } = ENDPOINTS.IDENTIFY;
+
+  const playerId = getExternalIdentifiersMapping(message);
+  const tags = {};
+
+  const traits = getFieldValueFromMessage(message, "traits");
+  const traitsKey = Object.keys(traits);
+  traitsKey.forEach(key => {
+    tags[key] = traits[key];
+  });
+  if (message.anonymousId) {
+    tags.anonymousId = message.anonymousId;
+  }
+  const payload = constructPayload(
+    message,
+    mappingConfig[ConfigCategory.IDENTIFY.name]
+  );
+  if (playerId) {
+    endpoint = `${endpoint}/${playerId}`;
+    payload.tags = tags;
+    payload.app_id = appId;
+    return responseBuilder(payload, endpoint);
+  }
+  populateDevicetype(message, payload);
+  const responseArray = [];
+  payload.tags = tags;
+  if (emailDeviceType) {
+    const emailDevicePayload = { ...payload };
+    emailDevicePayload.device_type = 11;
+    emailDevicePayload.identifier = getFieldValueFromMessage(message, "email");
+    responseArray.push(responseBuilder(emailDevicePayload, endpoint));
+  }
+  if (smsDeviceType) {
+    const smsDevicePayload = { ...payload };
+    smsDevicePayload.device_type = 14;
+    smsDevicePayload.identifier = getFieldValueFromMessage(message, "phone");
+    responseArray.push(responseBuilder(smsDevicePayload, endpoint));
+  }
+  if (payload.device_type) {
+    responseArray.push(responseBuilder(payload, endpoint));
+  }
+  return responseArray;
+};
+
+const trackResponseBuilder = (message, { Config }) => {
+  const { appId, eventAsTags, allowedProperties } = Config;
+  if (!appId) {
+    throw new CustomError(
+      "[OneSignal]: appId is required for creating/adding a device",
+      400
+    );
+  }
+  const event = get(message, "event");
+  let { endpoint } = ENDPOINTS.TRACK;
+  const externalUserId = getFieldValueFromMessage(message, "userId");
+  if (!externalUserId) {
+    throw new CustomError(
+      "[OneSignal]: userId is required for track events/updating a device",
+      400
+    );
+  }
+  endpoint = `${endpoint}/${appId}/users/${externalUserId}`;
+  const payload = {};
+  const tags = {};
+  if (event) {
+    tags[event] = true;
+  }
```

Comment:
We should throw an error if `event` is not defined.

Reply:
done